File: tree/java/BinarySearchTree.java
// Recursive

class Node {
    int data;
    Node left, right;

    public Node(int data) {
        this.data = data;
        left = right = null;
    }
}

public class BinarySearchTree {
    Node root;

    public int FindMinDepth() {

        return findMinDepth(root);
    }

    private int findMinDepth(Node root) {
        if (root == null) {
            return 0;
        }
        if (root.left == null && root.right == null) {
            return 1;
        }

        if (root.left == null) {
            return findMinDepth(root.right) + 1;
        }
        if (root.right == null) {
            return findMinDepth(root.left) + 1;
        }

        return Math.min(findMinDepth(root.left), findMinDepth(root.right)) + 1;
    }

    public static void main(String[] args) {
        BinarySearchTree tree = new BinarySearchTree();
        tree.root = new Node(1);
        tree.root.left = new Node(2);
        tree.root.right = new Node(3);
        tree.root.left.left = new Node(4);
        tree.root.left.right = new Node(5);

        System.out.println("the minimum depth of the tree is " + tree.FindMinDepth());
        /*
         * 1
         * / \
         * 2 3
         * / \
         * 4 5
         */

    }
}

File: tree/python/binary_search_tree.py
class Node:
    def __init__(self, key) -> None:
        self.key = key
        self.left = None
        self.right = None


def insert(node, key) -> Node:
    if node is None:
        return Node(key)
    if key > node.key:
        node.right = insert(node.right, key)
    else:
        node.left = insert(node.left, key)
    # return pointer
    return node


def search(node, key) -> Node:
    print("finding {}".format(key))


if __name__ == '__main__':
    root = None
    root = insert(root, 50)
    print('key:{}, left:{}, right:{}'.format(root.key, root.left, root.right))
    insert(root, 30)
    print('key:{}, left:{}, right:{}'.format(
        root.key, root.left.key, root.right))

    insert(root, 20)
    print('key:{}, left:{}, right:{}'.format(
        root.left.key, root.left.left.key, root.right))
    insert(root, 40)
    print('key:{}, left:{}, right:{}'.format(
        root.left.key, root.left.left.key, root.left.right.key))
    insert(root, 70)
    print('key:{}, left:{}, right:{}'.format(
        root.key, root.left.key, root.right.key))
    # Key to be found
    key = 6

    # Searching in a BST
    if search(root, key) is None:
        print(key, "not found")
    else:
        print(key, "found")
